flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceWithWatermarkExtractor.java

```diff
@@ -0,0 +1,359 @@
+                    // Just print all the data to confirm everything has arrived
+                    values.forEach(out::collect);
+                  }
+                });
+
+    try (CloseableIterator<RowData> resultIterator = windowed.collectAsync()) {
+      env.executeAsync("Iceberg Source Windowing Test");
+
+      // Write data so the windows containing test data are closed
+      dataAppender.appendToTable(ImmutableList.of(generateRecord(1500, "last-record")));
+      dataAppender.appendToTable(ImmutableList.of(generateRecord(1500, "last-record")));
+      dataAppender.appendToTable(ImmutableList.of(generateRecord(1500, "last-record")));
+
+      assertRecords(resultIterator, expectedRecords);
+    }
+  }
+
+  @Test
+  public void testThrottling() throws Exception {
+    GenericAppenderHelper dataAppender = appender();
+
+    // Generate records with the following pattern:
+    // - File 1 - Later records (Watermark 6000000)
+    //    - Split 1 - 2 records (100, "file_1-recordTs_100"), (103, "file_1-recordTs_103")
+    // - File 2 - First records (Watermark 0)
+    //    - Split 1 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    //    - Split 2 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    List<Record> batch;
+    batch =
+        ImmutableList.of(
+            generateRecord(100, "file_1-recordTs_100"), generateRecord(103, "file_1-recordTs_103"));
+    dataAppender.appendToTable(batch);
+
+    batch = Lists.newArrayListWithCapacity(100);
+    for (int i = 0; i < RECORD_NUM_FOR_2_SPLITS; ++i) {
+      batch.add(generateRecord(i % 5, "file_2-recordTs_" + i));
+    }
+
+    dataAppender.appendToTable(batch);
+
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.setParallelism(2);
+
+    DataStream<RowData> stream =
+        env.fromSource(
+            sourceBuilder()
+                .streaming(true)
+                .monitorInterval(Duration.ofMillis(10))
+                .streamingStartingStrategy(StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL)
+                .build(),
+            WatermarkStrategy.<RowData>noWatermarks()
+                .withWatermarkAlignment("iceberg", Duration.ofMinutes(20), Duration.ofMillis(10)),
+            SOURCE_NAME,
+            TypeInformation.of(RowData.class));
+
+    try (CloseableIterator<RowData> resultIterator = stream.collectAsync()) {
+      JobClient jobClient = env.executeAsync("Continuous Iceberg Source Failover Test");
+
+      // Check that the read the non-blocked data
+      // The first RECORD_NUM_FOR_2_SPLITS should be read
+      // 1 or more from the runaway reader should be arrived depending on thread scheduling
+      waitForRecords(resultIterator, RECORD_NUM_FOR_2_SPLITS + 1);
+
+      // Get the drift metric, wait for it to be created and reach the expected state
+      // (100 min - 20 min - 0 min)
```

Comment:
I am wondering if this can demonstrate throttling/alignment better
f1: 1 records with ts = 1 min
f2: 50 records with ts = 2 min
f3: 10 records with ts = 50 min
f4: 1 record with ts = 40 min

- setup: parallelism is 2. source discovery interval is 1 ms. watermark update interval is 1 ms. max drift is 20 mins
- add a map function that sleep 1 ms for every msg processed to slow down the processing.
- append f3 to the table after result collected f1-r1 (indicating f1 finished).
- append f4 to the table after result collected f2-r50 (indicating f2 finished).
- waiting for f3-r10. some of the f3 records should come after the watermark advanced to minute 40 with f2 finished and f4 arrived. this is an indication of alignment/throttling on f3. without alignment, f3 would have finished before f2.

we need to add a util method to wait for a specific record.

My concern is if this type of testing can be flaky.

Reply:
> can you help me understand why the drift should reach 80 minutes?
> 
> file 1 would advance reader 1 watermark to 100. file 2 would advance treader 2 watermark to 0. wouldn't drift be 100? why would it eventually come down to 80 without new splits?

Because the metrics are already calculate with the

Reply:
Exact testing is not working for throttling, as there is an inherently time sensitive element in global watermark propagation which happens asynchronously. So we can not be sure how many of the too early records will be propagated before the throttling kicks in, and the reader is stopped.

That is the reason why even Flink internal tests are depend on the the `WATERMARK_ALIGNMENT_DRIFT` metrics to check if the throttling is happened.

We follow the same pattern here:
- Add some new records, add some old records
- Verify that the old records are received fully
- Check the alignment metrics
- Add some more new records 
- Verify that all the records are received
- Verify the alignment metrics